File: DATASTRUCTURES/src/main/python/mylib/datastructures/trees/BST.py
from datastructures.nodes.TNode import TNode


class BST:
    def __init__(self):
        self.root = None

    def __init__(self, val):
        self.root = TNode(val, 0, None, None, None)

    def __init__(self, obj):
        self.root = obj

    def set_root(self, node):
        self.root = node

    def get_root(self):
        return self.root
    def insert(self, val):
        if self.root is None:
            self.root = TNode(val, 0, None, None, None)
        else:
            self._insert(val, self.root)

    def _insert(self, val, node):
        if val < node.data:
            if node.left is None:
                node.left = TNode(val, 0, node, None, None)
                self._update_balance(node, 1)
            else:
                self._insert(val, node.left)
        else:
            if node.right is None:
                node.right = TNode(val, 0, node, None, None)
                self._update_balance(node, -1)
            else:
                self._insert(val, node.right)

    def insert_node(self, node):
        if self.root is None:
            self.root = node
        else:
            self._insert_node(node, self.root)

    def _insert_node(self, node, current):
        if node.data < current.data:
            if current.left is None:
                current.left = node
                node.parent = current
                self._update_balance(current, 1)
            else:
                self._insert_node(node, current.left)
        else:
            if current.right is None:
                current.right = node
                node.parent = current
                self._update_balance(current, -1)
            else:
                self._insert_node(node, current.right)

    def Delete(self, val):
        current = self.root
        parent = None

        # Search for the node with the given value
        while current != None and current.data != val:
            parent = current
            if val < current.data:
                current = current.left
            else:
                current = current.right

        # If node not found, print a statement and return
        if current == None:
            print(f"Value {val} not found in the tree.")
            return

        # If node has no children, delete it and update parent's child pointer
        if current.left == None and current.right == None:
            if current == self.root:
                self.root = None
            elif current == parent.left:
                parent.left = None
            else:
                parent.right = None
            del current

        # If node has one child, bypass it and update parent's child pointer
        elif current.left == None or current.right == None:
            child = current.left if current.left != None else current.right
            if current == self.root:
                self.root = child
            elif current == parent.left:
                parent.left = child
            else:
                parent.right = child
            del current

        # If node has two children, find the successor and replace the node's data
        else:
            successor = current.right
            while successor.left != None:
                successor = successor.left
            current.data = successor.data
            self.Delete(successor.data)

    def Search(self, val):
        current = self.root
        while current != None:
            if current.data == val:
                return current
            elif val < current.data:
                current = current.left
            else:
                current = current.right
        return None

    def printInOrder(self):
        self._printInOrderHelper(self.root)
        print()

    def _printInOrderHelper(self, node):
        if node != None:
            self._printInOrderHelper(node.left)
            print(node.data, end=' ')
            self._printInOrderHelper(node.right)
    def printBF(self):
        pass

File: src/main/java/mylib/datastructures/trees/BST.java
package main.java.mylib.datastructures.trees;

import main.java.mylib.datastructures.nodes.TNode;

import java.util.LinkedList;
import java.util.Queue;

/**
 * Binary Search Tree class
 */
public class BST {
    private TNode root;

    /**
     * Default constructor
     */
    public BST() {
        root = null;
    }

    /**
     * Overload constructor
     * 
     * @param val the data of the root node
     */
    public BST(int val) {
        this.root = new TNode(val, 0, null, null, null);
    }

    /**
     * Overload constructor
     * 
     * @param obj the root node
     */
    public BST(TNode obj) {
        this.root = obj;
    }

    /**
     * setter for the root
     * 
     * @param root the root node
     */
    public void setRoot(TNode root) {
        this.root = root;
    }

    /**
     * getter for the root
     * 
     * @return the root node
     */
    public TNode getRoot() {
        return root;
    }

    /**
     * insert a data into the tree
     * 
     * @param val the data to be inserted
     */
    public void insert(int val) {
        TNode newNode = new TNode(val, 0, null, null, null);
        if (root == null) {
            root = newNode;
        } else {
            insert(root, newNode);
        }
    }

    /**
     * insert a node into the tree
     * 
     * @param node    the root of the tree
     * @param newNode the node to be inserted
     */
    private void insert(TNode node, TNode newNode) {
        if (newNode.getData() < node.getData()) {
            if (node.getLeft() == null) {
                node.setLeft(newNode);
            } else {
                insert(node.getLeft(), newNode);
            }
        } else {
            if (node.getRight() == null) {
                node.setRight(newNode);
            } else {
                insert(node.getRight(), newNode);
            }
        }
    }

    /**
     * insert a node into the tree
     * 
     * @param node the node to be inserted
     */
    public void insert(TNode node) {
        if (root == null) {
            root = node;
        } else {
            insert(root, node);
        }
    }

    /**
     * delete a node from the tree
     * 
     * @param val the value of the node to be deleted
     */
    public void delete(int val) {
        root = delete(root, val);
    }

    /**
     * delete a node from the tree
     * 
     * @param node the root of the tree
     * @param val  the value of the node to be deleted
     * @return the node that is deleted
     */
    private TNode delete(TNode node, int val) {
        if (node == null) {
            System.out.println("Value " + val + " is not in the tree.");
            return null;
        }
        if (val < node.getData()) {
            node.setLeft(delete(node.getLeft(), val));
        } else if (val > node.getData()) {
            node.setRight(delete(node.getRight(), val));
        } else {
            if (node.getLeft() == null && node.getRight() == null) {
                node = null;
            } else if (node.getLeft() == null) {
                node = node.getRight();
            } else if (node.getRight() == null) {
                node = node.getLeft();
            } else {
                TNode temp = findMin(node.getRight());
                node.setData(temp.getData());
                node.setRight(delete(node.getRight(), temp.getData()));
            }
        }
        return node;
    }

    /**
     * find the minimum node in the tree
     * 
     * @param node the root of the tree
     * @return the minimum node
     */
    private TNode findMin(TNode node) {
        while (node.getLeft() != null) {
            node = node.getLeft();
        }
        return node;
    }

    /**
     * search a node in the tree
     * 
     * @param val the value of the node to be searched
     * @return the node that is searched
     */
    public TNode search(int val) {
        return search(root, val);
    }

    /**
     * search a node in the tree
     * 
     * @param node the root of the tree
     * @param val  the value of the node to be searched
     * @return the node that is searched
     */
    private TNode search(TNode node, int val) {
        if (node == null || node.getData() == val) {
            return node;
        }
        if (val < node.getData()) {
            return search(node.getLeft(), val);
        } else {
            return search(node.getRight(), val);
        }
    }

    /**
     * print the tree in order
     */
    public void printInOrder() {
        printInOrder(root);
        System.out.println();
    }

    /**
     * print the tree in order
     * 
     * @param node the root of the tree
     */
    private void printInOrder(TNode node) {
        if (node != null) {
            printInOrder(node.getLeft());
            System.out.print(node.getData() + " ");
            printInOrder(node.getRight());
        }
    }

    /**
     * print the tree in BF
     */
    public void printBF() {
        if (root == null) {
            return;
        }

        Queue<TNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                TNode current = queue.remove();
                System.out.print(current.getData() + " ");

                if (current.getLeft() != null) {
                    queue.add(current.getLeft());
                }

                if (current.getRight() != null) {
                    queue.add(current.getRight());
                }
            }

            System.out.println(); // Print newline after each level
        }
    }
}
